File: C#/CompilerStudy/Lesson1_AntlrLang/JSMinus/AST.cs

using System.Collections.Generic;

// ExprNode
public interface IExprNodeVisitor {
    void visit(ExprNode_ConstNumber node);
    void visit(ExprNode_ConstString node);
    void visit(ExprNode_ID node);
    void visit(ExprNode_ArrayConstructor node);
    void visit(ExprNode_UnaryOp node);
    void visit(ExprNode_BinaryOp node);
    void visit(ExprNode_IndexOf node);
    void visit(ExprNode_Call node);
}

public abstract class ExprNode {
    protected ExprNode(int line) { Line = line; }
    public int Line { get; private set; }
    public abstract void acceptVisitor(IExprNodeVisitor v);
}

public class ExprNode_ConstNumber : ExprNode {
    public ExprNode_ConstNumber(int line, double number) : base(line) { Number = number; }
    public double Number { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_ConstString : ExprNode {
    public ExprNode_ConstString(int line, string str) : base(line) { Str = str;}
    public string Str { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_ID : ExprNode { 
    public ExprNode_ID(int line, string name) : base(line) { Name = name;  }
    public string Name { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_ArrayConstructor : ExprNode {
    public ExprNode_ArrayConstructor(int line, List<ExprNode> exprs) : base(line) { Exprs = exprs; }
    public List<ExprNode> Exprs { get; private set; } 
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_UnaryOp : ExprNode {
    public ExprNode_UnaryOp(int line, string op, ExprNode expr) : base(line) { Op = op; Expr = expr; }
    public string Op { get; private set; }
    public ExprNode Expr { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_BinaryOp : ExprNode {
    public ExprNode_BinaryOp(int line, string op, ExprNode lexpr, ExprNode rexpr) : base(line) { Op = op; Lexpr = lexpr; Rexpr = rexpr; }
    public string Op { get; private set; }
    public ExprNode Lexpr { get; private set; }
    public ExprNode Rexpr { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_IndexOf : ExprNode {
    public ExprNode_IndexOf(int line, ExprNode arrayExpr, ExprNode indexExpr):base(line) { ArrayExpr = arrayExpr; IndexExpr = indexExpr; }
    public ExprNode ArrayExpr { get; private set; }
    public ExprNode IndexExpr { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
public class ExprNode_Call : ExprNode {
    public ExprNode_Call(int line, ExprNode funcExpr, List<ExprNode> paramExprs) : base(line) { FuncExpr = funcExpr; ParamExprs = paramExprs; }
    public ExprNode FuncExpr { get; private set; }
    public List<ExprNode> ParamExprs { get; private set; }
    public override void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}

// StmtNode
public interface IStmtNodeVisitor {
    void visit(StmtNode_DeclareLocal node);
    void visit(StmtNode_DeclareArg node);
    void visit(StmtNode_Block node);
    void visit(StmtNode_Stmts node);
    void visit(StmtNode_Assign node);
    void visit(StmtNode_Call node);
    void visit(StmtNode_IfElse node);
    void visit(StmtNode_For node);
    void visit(StmtNode_Continue node);
    void visit(StmtNode_Break node);
    void visit(StmtNode_Return node);
}

public abstract class StmtNode {
    public abstract void acceptVisitor(IStmtNodeVisitor v);
    protected StmtNode(int line) { Line = line;  }
    public int Line { get; set; }
}

public class StmtNode_DeclareLocal : StmtNode {
    public StmtNode_DeclareLocal(int line, string name) : base(line) { Name = name;  }
    public string Name { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_DeclareArg : StmtNode {
    public StmtNode_DeclareArg(int line, string name): base(line) { Name = name; }
    public string Name { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Block: StmtNode {
    public StmtNode_Block(int line) : base(line) { Stmts = new List<StmtNode>();}
    public List<StmtNode> Stmts { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Stmts: StmtNode {
    public StmtNode_Stmts(int line) : base(line) { Stmts = new List<StmtNode>();}
    public List<StmtNode> Stmts { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Assign: StmtNode {
    public StmtNode_Assign(int line, ExprNode lexpr, ExprNode rexpr) : base(line) { Lexpr = lexpr; Rexpr = rexpr; }
    public ExprNode Lexpr { get; private set; }
    public ExprNode Rexpr { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Call: StmtNode {
    public StmtNode_Call(int line, ExprNode_Call callExpr):base(line) { CallExpr = callExpr; }
    public ExprNode_Call CallExpr { get;private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_IfElse : StmtNode {
    public StmtNode_IfElse(int line, ExprNode expr, StmtNode ifStmt, StmtNode elseStmt) : base(line) { Expr = expr; IfStmt = ifStmt; ElseStmt = elseStmt; }
    public ExprNode Expr { get; private set; }
    public StmtNode IfStmt { get; private set; }
    public StmtNode ElseStmt { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_For : StmtNode {
    public StmtNode_For(int line, StmtNode first, ExprNode second, StmtNode third, StmtNode body) : base(line) { First = first; Second = second; Third = third; Body = body; }
    public StmtNode First { get; private set; }
    public ExprNode Second { get; private set; }
    public StmtNode Third { get; private set; }
    public StmtNode Body { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Continue : StmtNode {
    public StmtNode_Continue(int line): base(line) { }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Break : StmtNode {
    public StmtNode_Break(int line) : base(line) { }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
public class StmtNode_Return : StmtNode {
    public StmtNode_Return(int line, ExprNode expr) : base(line) { Expr = expr; }
    public ExprNode Expr { get; private set; }
    public override void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}

File: Java/CompilerStudy/Lesson1_AntlrLang/AST.java
import java.util.*;

// ExprNode
interface IExprNodeVisitor {
    void visit(ExprNode_ConstNumber node);
    void visit(ExprNode_ConstString node);
    void visit(ExprNode_ID node);
    void visit(ExprNode_ArrayConstructor node);
    void visit(ExprNode_UnaryOp node);
    void visit(ExprNode_BinaryOp node);
    void visit(ExprNode_IndexOf node);
    void visit(ExprNode_Call node);
}

abstract class ExprNode {
    protected ExprNode(int _line) { line = _line; }
    public int line;
    public abstract void acceptVisitor(IExprNodeVisitor v);
}

class ExprNode_ConstNumber extends ExprNode {
    public ExprNode_ConstNumber(int _line, double _number) { super(_line); number = _number; }
    public double number;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_ConstString extends ExprNode {
    public ExprNode_ConstString(int _line, String _str) { super(_line); str = _str;}
    public String str;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_ID extends ExprNode { 
    public ExprNode_ID(int _line, String _name) { super(_line); name = _name;  }
    public String name;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_ArrayConstructor extends ExprNode {
    public ExprNode_ArrayConstructor(int _line, ArrayList<ExprNode> _exprs) { super(_line); exprs = _exprs; }
    public ArrayList<ExprNode> exprs;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_UnaryOp extends ExprNode {
    public ExprNode_UnaryOp(int _line, String _op, ExprNode _expr) { super(_line); op = _op; expr = _expr; }
    public String op;
    public ExprNode expr;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_BinaryOp extends ExprNode {
    public ExprNode_BinaryOp(int _line, String _op, ExprNode _lexpr, ExprNode _rexpr) { super(_line); op = _op; lexpr = _lexpr; rexpr = _rexpr; }
    public String op;
    public ExprNode lexpr, rexpr;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_IndexOf extends ExprNode {
    public ExprNode_IndexOf(int _line, ExprNode _arrayExpr, ExprNode _indexExpr) { super(_line); arrayExpr = _arrayExpr; indexExpr = _indexExpr; }
    public ExprNode arrayExpr,  indexExpr;
    @Override
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}
class ExprNode_Call extends ExprNode {
    public ExprNode_Call(int _line, ExprNode _funcExpr, ArrayList<ExprNode> _paramExprs) { super(_line); funcExpr = _funcExpr; paramExprs = _paramExprs; }
    public ExprNode funcExpr;
    public ArrayList<ExprNode> paramExprs;
    @Override 
    public void acceptVisitor(IExprNodeVisitor v) { v.visit(this); }
}

// StmtNode
interface IStmtNodeVisitor {
    void visit(StmtNode_DeclareLocal node);
    void visit(StmtNode_DeclareArg node);
    void visit(StmtNode_Block node);
    void visit(StmtNode_Stmts node);
    void visit(StmtNode_Assign node);
    void visit(StmtNode_Call node);
    void visit(StmtNode_IfElse node);
    void visit(StmtNode_For node);
    void visit(StmtNode_Continue node);
    void visit(StmtNode_Break node);
    void visit(StmtNode_Return node);
}

abstract class StmtNode {
    public abstract void acceptVisitor(IStmtNodeVisitor v);
    protected StmtNode(int _line) { line = _line;  }
    public int line;
}

class StmtNode_DeclareLocal extends StmtNode {
    public StmtNode_DeclareLocal(int _line, String _name) { super(_line); name = _name;  }
    public String name;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_DeclareArg extends StmtNode {
    public StmtNode_DeclareArg(int _line, String _name) { super(_line); name = _name; }
    public String name;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Block extends StmtNode {
    public StmtNode_Block(int _line) { super(_line); stmts = new ArrayList<StmtNode>();}
    public ArrayList<StmtNode> stmts;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Stmts extends StmtNode {
    public StmtNode_Stmts(int _line) { super(_line); stmts = new ArrayList<StmtNode>();}
    public ArrayList<StmtNode> stmts;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Assign extends StmtNode {
    public StmtNode_Assign(int _line, ExprNode _lexpr, ExprNode _rexpr) { super(_line); lexpr = _lexpr; rexpr = _rexpr; }
    public ExprNode lexpr, rexpr;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Call extends StmtNode {
    public StmtNode_Call(int _line, ExprNode_Call _callExpr) { super(_line); callExpr = _callExpr; }
    public ExprNode_Call callExpr;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_IfElse extends StmtNode {
    public StmtNode_IfElse(int _line, ExprNode _expr, StmtNode _ifStmt, StmtNode _elseStmt) { super(_line); expr = _expr; ifStmt = _ifStmt; elseStmt = _elseStmt; }
    public ExprNode expr;
    public StmtNode ifStmt, elseStmt;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_For extends StmtNode {
    public StmtNode_For(int _line, StmtNode _first, ExprNode _second, StmtNode _third, StmtNode _body) { 
        super(_line);
        first = _first; second = _second; third = _third; body = _body; 
    }
    public StmtNode first, third, body;
    public ExprNode second;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Continue extends StmtNode {
    public StmtNode_Continue(int _line) { super(_line);}
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Break extends StmtNode {
    public StmtNode_Break(int _line) { super(_line); }
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
class StmtNode_Return extends StmtNode {
    public StmtNode_Return(int _line, ExprNode _expr)  { super(_line); expr = _expr; }
    public ExprNode expr;
    @Override
    public void acceptVisitor(IStmtNodeVisitor v) { v.visit(this); }
}
